flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlBinlogSplitAssigner.java

```diff
@@ -127,7 +154,78 @@ public void close() {}
 
     // ------------------------------------------------------------------------------------------
 
-    private MySqlBinlogSplit createBinlogSplit() {
+    private MySqlBinlogSplit createBinlogSplitFromTimestamp() {
+        BinlogOffset binlogOffset = createBinlogOffsetFromTimestamp();
+        return new MySqlBinlogSplit(
+                BINLOG_SPLIT_ID,
+                binlogOffset,
+                BinlogOffset.NO_STOPPING_OFFSET,
+                new ArrayList<>(),
+                new HashMap<>(),
+                0);
+    }
+
+    private BinlogOffset createBinlogOffsetFromTimestamp() {
+        List<String> databaseList = sourceConfig.getDatabaseList();
+        Preconditions.checkArgument(
+                databaseList.size() == 1, "startup by timestamp only support 1 mysql database");
+
+        Long startupTimestampMillis = this.startupOptions.startupTimestampMillis;
+        Preconditions.checkArgument(
+                startupTimestampMillis < System.currentTimeMillis(),
+                "startup timestamp should be small than now .");
+
+        final long beginTimestampMills = System.currentTimeMillis();
+        LOG.info("Begin to find binlog offset by timestamp {}", startupTimestampMillis);
+        try (JdbcConnection jdbc = DebeziumUtils.openJdbcConnection(sourceConfig)) {
+            BinlogOffset maxBinlogOffset = currentBinlogOffset(jdbc);
+            String binlogFile = maxBinlogOffset.getFilename();
+            BinlogOffset seekBinlogOffset = null;
+            boolean shouldBreak = false;
+            while (!shouldBreak) {
+                BinlogOffset binlogOffset =
+                        DebeziumUtils.seekBinlogOffsetByTimestamp(
+                                binlogFile, maxBinlogOffset, sourceConfig, startupTimestampMillis);
+
+                // Can not find the binlog offset if the binlog offset's timestamp is 0 .
+                // see SeekBinlogTimestampListener to find more detail
+                if (binlogOffset == null || binlogOffset.getTimestamp() == 0) {
```

Comment:
Is there a case where binlogTimestamp is 0 while binlogOffset has been instantiated?